### DISCUSSION TITLE
一次zig踩坑记录

### DISCUSSION THREAD

**Discussion by zhuyadong** [zigcc/forum#78] [General]
先上代码：
```zig
const std = @import("std");
const FieldEnum = std.meta.FieldEnum;
const FieldType = std.meta.FieldType;
fn fieldIsPtr(comptime T: type, comptime field: FieldEnum(T)) bool {
    return (struct {
        pub const ret = blk: {
            const fieldType = FieldType(T, field);
            break :blk (@typeInfo(fieldType) == .Pointer);
        };
    }).ret;
}

fn fieldPtrType(comptime T: type, comptime field: ?FieldEnum(T)) type {
    return (struct {
        pub const ret = blk: {
            if (field) |fld| {
                switch (fieldIsPtr(T, fld)) {
                    true => break :blk FieldType(T, fld),
                    false => break :blk *FieldType(T, fld),
                }
            }
            break :blk *T;
        };
    }).ret;
}

fn fieldPtr(ptr: anytype, comptime field: ?FieldEnum(@TypeOf(ptr.*))) fieldPtrType(@TypeOf(ptr.*), field) {
    const T = @TypeOf(ptr.*);
    if (field != null) {
        if (fieldIsPtr(T, field.?)) {
            return @field(ptr.*, @tagName(field.?));
        } else {
            return @ptrFromInt(@intFromPtr(ptr) + @offsetOf(T, @tagName(field.?)));
        }
    } else {
        return ptr;
    }
}

const Info = struct {
    name: []const u8 = "aeix",
};

const Data = struct {
    age: u32 = 0,
    name: *u32 = undefined,
    info: Info = .{},
};

pub fn main() void {
    var data = Data{};
    const pdata = &data;
    std.debug.assert(fieldPtr(pdata, .age) == &pdata.age);
    std.debug.assert(fieldPtr(pdata, .name) == pdata.name);
}
```
这段代码主要想测试的是`fieldPtr`这个函数，这个函数作用就是想把第一个参数的由第二个参数指定的field返回，并保证返回的是一个指针。
上面代码在zig 0.12.0-dev.1861+412999621 下编译会失败，报错：
```
src\main.zig:31:20: error: expected type '*u32', found 'u32'
            return @field(ptr.*, @tagName(field.?));
```
修复方法很简单，*只要把函数`fieldIsPtr`改成inline的就可以了*，至于原因，本人目前不是很清楚，所以留下贴子希望有人能解答。

**Comment by xiamuguizhi** [comment]
回答自GTP4：

在 Zig 中，`comptime` 表示在编译时进行计算。错误信息：

```
src\main.zig:31:20: error: expected type '*u32', found 'u32'
            return @field(ptr.*, @tagName(field.?));
```

意味着 `@field` 函数期望返回一个指针类型 (`*u32`)，但实际上返回的是非指针类型 (`u32`)。这是因为 `@field` 函数访问结构体的字段并返回该字段的值，但在这种情况下，字段 `name` 已经是一个指针类型，所以直接返回 `name` 字段的值（即指针）是正确的。

当你将 `fieldIsPtr` 函数改为 `inline` 函数时，你强制编译器在每个调用点内联该函数，这意味着编译器在处理 `fieldIsPtr` 函数的调用时会计算出确切的返回值，而不是将其作为可运行时计算的表达式。这样做可以确保在编译时就能知道 `fieldIsPtr` 的返回值，所以编译器可以正确处理 `@field` 或 `@ptrFromInt` 的调用。

在你的代码中，`fieldIsPtr` 函数检查一个字段是否是指针类型。如果不使用 `inline`，编译器可能无法在编译时确定 `fieldIsPtr` 的返回值，因为它是在一个运行时的块（`blk`）中计算的。这意味着编译器不能确定 `@field` 应该返回指针还是值，因此会报错。

使用 `inline` 关键字后，编译器在编译时就会展开 `fieldIsPtr` 函数，这样就可以在编译时确定每个字段是否为指针，并据此决定使用 `@field` 还是 `@ptrFromInt`。

这就是为什么将 `fieldIsPtr` 改为 `inline` 可以修复这个问题的原因。通过内联，编译器能够在编译时期解析出所有必要的类型信息，从而正确处理类型转换和函数返回值。

**Reply by Clownsw** [reply]
这个错误有点误导人, 即使没有加 `inline` 解出来的的是 `u32`, 那么 `return &@field(ptr.*, @tagName(field.?)); // 当然这段代码是有问题的, 只是确认可以不可把类型切换到 *u32` 就变成了 `**u32`, 感觉zig在错误提示这块还是不太行
